File: Python/0053_maximum_subarray/0053_maximum_subarray.py
class Solution:
    def maxSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        maximum = nums[0]
        dp = []
        dp.append(nums[0])
        
        for i in range(1, len(nums)):
            if dp[i-1] > 0:
                dp.append(nums[i] + dp[i-1])
            else:
                dp.append(nums[i])
            maximum = max(dp[i], maximum)

        return maximum

File: Cplusplus/0053_maximum_subarray/0053_maximum_subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maximum = nums[0];
        int* dp = new int[nums.size()];
        dp[0] = nums[0];
        
        for (int i = 1; i < nums.size(); i++)
        {
            dp[i] = dp[i-1] > 0 ? dp[i-1] + nums[i] : nums[i];
            maximum = max(maximum, dp[i]);
        }
        delete[] dp;
        
        return maximum;
    }
};

